pinyin-search.el

```diff
@@ -91,24 +102,39 @@ see URL `https://github.com/redguardtoo/find-by-pinyin-dired'.")
     ;; Return default function
     (isearch-search-fun-default)))
 
-(defun isearch-turn-off-pinyin ()
-  (setq isearch-pinyin nil)
-  (setq mode-line-format (delete isearch-pinyin-mode-line-indicate
-                                 mode-line-format))
+(defun isearch-update-pinyin-indicator ()
+  (if isearch-pinyin
+      (add-to-list 'mode-line-format isearch-pinyin-mode-line-indicate)
+    (setq mode-line-format
+          (delete isearch-pinyin-mode-line-indicate mode-line-format)))
   (force-mode-line-update))
 
+(defun isearch-set-pinyin-state ()
+  ;; Only when users cancel isearch or exit isearch normally with
+  ;; `isearch-exit' should the state of pinyin search be changed.
```

Comment:
Functions in `isearch-mode-end-hook` are called by `isearch-done`, and the key point here is: `isearch-done` is not only called when users exit isearch normally or cancel it, but also called when entering isearch edit mode. So it seems we have to distinguish different contexts in case of the problem that pinyin search is turned off automatically when editing the string.